### DISCUSSION TITLE
Help with simple usage

### DISCUSSION THREAD

**Discussion by LGMAM** [alecthomas/participle#415] [Q&A] [answered]
Hi all,

Just starting to use this library to parse a custom file format, quite similiar to INI files.

I am not able to get right the most simple example.

This is the file contents

```
==abc
zzz
```
And this is the parser

```
var lexerDefinition = lexer.MustSimple([]lexer.SimpleRule{
	{Name: "Ident", Pattern: `[a-zA-Z][a-zA-Z_\d]*`},
})

var Parser = participle.MustBuild[Matadown](
	participle.Lexer(lexerDefinition),
)

type Matadown struct {
	Sections []*Section `@@*`
}

type Section struct {
	Title   string   `"==" @Ident`
	Content *Content `@@`
}

type Content struct {
	String string `@Ident`
}
```

Getting this error

`1:1: invalid input text "==abc\nzzz\n\n"`

Any hint ?

Thanks

**Comment by alecthomas** [comment, marked as answer]
Traditional lexer/parsers like Participle operate in two independent phases: the lexer extracts named tokens from the raw text, then the parser uses those tokens to construct an AST. Your lexer only has tokens for "Ident", so when it encounters "==" it fails. You'll need to define rules for any possible text you might encounter. Also note that the rules are tried in order, first match wins.
